File: app/src/main/java/amal/global/amal/PhotoStorage.java
package amal.global.amal;

import android.content.Context;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;

/**
 * Created by soroushkhanlou on 11/13/17.
 */

public class PhotoStorage {

    Context context;

    PhotoStorage(Context context) {
        this.context = context;
    }

    void savePhotoLocally(byte[] bytes) {
        final File dir = new File(context.getFilesDir() + "/images/");
        dir.mkdirs();
        int maxImageID = -1;
        for (File file : dir.listFiles()) {
            String filename = file.getName();
            String IDString = FilenameUtils.removeExtension(filename);
            int id = Integer.parseInt(IDString);
            if (id > maxImageID) {
                maxImageID = id;
            }
        }
        final File file = new File(dir + String.format("/%d.jpg", maxImageID+1));
        OutputStream output = null;
        try {
            save(bytes, file);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void save(byte[] bytes, File file) throws IOException {
        OutputStream output = null;
        try {
            output = new FileOutputStream(file);
            output.write(bytes);
        } finally {
            if (null != output) {
                output.close();
            }
        }
    }

    public ArrayList<Image> fetchImages() {
        final File dir = new File(context.getFilesDir() + "/images/");
        dir.mkdirs();
        File[] files = dir.listFiles();
        ArrayList<Image> images = new ArrayList<Image>();
        for (File file : files) {
            String filename = file.getName();
            String fileExtension = FilenameUtils.getExtension(filename);
            String IDString = FilenameUtils.removeExtension(filename);
            if (fileExtension.equals("jpg") || fileExtension.equals("jpeg")) {
                images.add(new Image(file.getAbsolutePath()));
            }
        }
        return images;
    }

}

File: app/src/main/java/amal/global/amal/PhotoStorage.kt
package amal.global.amal

import android.content.Context
import android.util.Log
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

class PhotoStorage internal constructor(internal var context: Context) {

    @Throws(IOException::class)
    public fun savePhotoLocally(bytes: ByteArray, metadata: Metadata) {
        val dir = File(context.filesDir.toString() + "/images/")
        dir.mkdirs()

        val maxImageID = dir.listFiles()
                .mapNotNull { file ->  file.nameWithoutExtension.toIntOrNull() }
                .maxOrNull() ?: 0

        val newImageID = maxImageID + 1
        val imageFile = File(dir.toString() + "/$newImageID.jpg")
        val settingsFile = File(dir.toString() + "/$newImageID.json")
        try {
            save(bytes, imageFile)
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        settingsFile.writeText(Metadata.jsonAdapter.toJson(metadata))
    }

    @Throws(IOException::class)
    private fun save(bytes: ByteArray, file: File) {
        var output: OutputStream? = null
        try {
            output = FileOutputStream(file)
            output.write(bytes)
        } finally {
            output?.close()
        }
    }

    public fun deleteImage(imagePath: String, settingsPath: String) {
        Log.d("PhotoStorage","delete image was called")
        try {
            //delete photo jpeg file
            val imageFile = File(imagePath)
            imageFile.delete()
            //delete affiliated json file - no idea if this will work - just trying easy first
            val metaDataFile = File(settingsPath)
            metaDataFile.delete()
        } catch(e:Exception) {
           Log.e(e.toString(),"Error deleting photo")
        }
            //TODO should probably add some error protection at some point - also may eventually add to WorkManager
    }

    public fun fetchImages(): List<LocalImage> {
        val dir = File(context.filesDir.toString() + "/images/")
        dir.mkdirs()
        return dir
                .listFiles()
                .filter { file ->
                    file.extension == "jpg" || file.extension == "jpeg"
                }
                .map { file ->
                    LocalImage.convenience(file.absolutePath, file.absolutePath.replaceAfterLast(".", "json"))
                }
                .sortedBy { image ->
                    image.date
                }
                .reversed()
    }

}
